parsl/tests/test_python_apps/test_memoize_plugin.py

```diff
@@ -0,0 +1,53 @@
+import argparse
+
+import pytest
+
+import parsl
+from parsl.app.app import python_app
+from parsl.config import Config
+from parsl.dataflow.memoization import BasicMemoizer
+from parsl.dataflow.taskrecord import TaskRecord
+
+
+class DontReuseSevenMemoizer(BasicMemoizer):
+    def check_memo(self, task_record: TaskRecord):
+        if task_record['args'][0] == 7:
+            return None  # we didn't find a suitable memo record...
+        else:
+            return super().check_memo(task_record)
+
+
+def local_config():
+    return Config(memoizer=DontReuseSevenMemoizer())
+
+
+@python_app(cache=True)
+def random_uuid(x, cache=True):
+    import uuid
+    return str(uuid.uuid4())
+
+
+@pytest.mark.local
+def test_python_memoization(n=10):
+    """Testing python memoization disable
+    """
+
+    # TODO: this .result() needs to be here, not in the loop
+    # because otherwise we race to complete... and then
+    # we might sometimes get a memoization before the loop
+    # and sometimes not...
```

Comment:
Is this generally something that users should keep in mind when working with caching and memoization in Parsl? Namely that there could be race conditions that crop up for functions that are in a loop and return very quickly?

Reply:
This will happen if you launch two apps with the same parameters at "the same time", where the same time means without waiting for one of them to complete:

1) invoke my_app(7)

there's no checkpoint/memo so we submit it to be executed, as task 1

2) invoke my_app(7)

there's no checkpoint/memo so we submit it to be executed, as task 2

3) task 1 completes and its result is memoized as my_app 7

4) task 2 completes and its result is memozied as my_app 7, replacing the result from step 3.



Maybe possible to implement to avoid this (again low priority for me) is more like Haskell thunks:

1) invoke my_app(7)

there's no checkpoint/memo so we submit it to be executed, as task 1. we memoise its unpopulated future as my_app 7.

2) invoke my_app(7)

there's a memo future for it (with no result yet) - so use that for the result

3) task 1 completes, populating its result future.
This population of result future causes the 2nd invocation memo future to be populated too, and that completes.

so my_app is only run once.

Reply:
Oh, ok. I see what you mean. From a selfish perspective, I don't think the extra effort of memoizing a future is necessary for my work. 

Thanks a bunch for the explanation.